lib/internal/test_runner/test.js

```diff
@@ -639,12 +647,17 @@ class Test extends AsyncResource {
         return;
       }
 
-      await afterEach();
-      await after();
       this.pass();
+      try {
+        await afterEach();
+        await after();
+      } catch (err) {
+        // If one of the after hooks has thrown unset endTime so that the
+        // catch below can do its cancel/fail logic.
+        this.endTime = null;
+        throw err;
```

Comment:
Won't this result in `afterEach` and `after` being called twice if `after` throws/returns a rejected promise?

Reply:
I thought so too, but I was thinking about before my changes. It was doing the same as this before which is why I'm re-throwing it. Remember that the previous version called after/afterEach then pass so endTime wasn't set before either.
I have checked it isn't called twice before or after my changes but I can't confidently say why that's the case.
I ran this fragment to check ```js
const test = require('node:test');

test.it('test', () => { });

test.afterEach(ctx => {
  console.log('afterEach', ctx.name, ctx.passed, ctx.error);
  throw new Error('err');
});
```

Reply:
they are both wrapped with `runOnce` for that exact reason:
https://github.com/nodejs/node/blob/32ce1834c03a38397db2cf42137b9376f110733e/lib/internal/test_runner/test.js#L489
https://github.com/nodejs/node/blob/32ce1834c03a38397db2cf42137b9376f110733e/lib/internal/test_runner/test.js#L596-L600

Reply:
Thanks, that's reassuring to know. I didn't think to look where they were created